src/github.com/matrix-org/dendrite/clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,184 @@
+	if req.Method != http.MethodGet {
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("Bad method"),
+		}
+	}
+	mtag := newMTag()
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	dataByte, _ := json.Marshal(data)
+	err = json.Unmarshal(dataByte, &mtag)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: mtag,
+	}
+}
+
+// PutTag implements PUT /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func PutTag(req *http.Request, accountDB *accounts.Database, userID string, roomID string, tag string) util.JSONResponse {
+	if req.Method != http.MethodPut {
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("Bad method"),
+		}
+	}
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	//Check for existing entries of tags for this ROOM ID and localpart
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	mtag := newMTag()
+	var properties common.TagProperties
+
+	if reqErr := httputil.UnmarshalJSONRequest(req, &properties); reqErr != nil {
+		return *reqErr
+	}
+
+	if len(data) > 0 {
+		dataByte, _ := json.Marshal(data)
+		err = json.Unmarshal(dataByte, &mtag)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+
+	mtag.Tags[tag] = properties
+	newTagData, _ := json.Marshal(mtag)
+
+	if err := accountDB.SaveAccountData(
+		req.Context(), localpart, roomID, "m.tag", string(newTagData),
+	); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func DeleteTag(req *http.Request, accountDB *accounts.Database, userID string, roomID string, tag string) util.JSONResponse {
+
+	if req.Method != http.MethodDelete {
+		return util.JSONResponse{
+			Code: http.StatusMethodNotAllowed,
+			JSON: jsonerror.NotFound("Bad method"),
+		}
+	}
+
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "m.tag",
+	)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	mtag := newMTag()
+
+	if len(data) > 0 {
+		dataByte, _ := json.Marshal(data)
+		err = json.Unmarshal(dataByte, &mtag)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	} else {
+		//Error indicating there is no Tag data
```

Comment:
For reasons I'm not privy to, it seems that the spec [defines](https://matrix.org/docs/spec/client_server/r0.4.0.html#delete-matrix-client-r0-user-userid-rooms-roomid-tags-tag) (and indeed Synapse follows) the fact that even if a tag does not already exist when you request a DELETE, a 200 OK is returned regardless.

This may be for privacy reasons, I'm not sure.

Either way, we shouldn't be returning an error here if the tag does not exist.

Reply:
Sure, changes will be done!

Reply:
Was this actually resolved?

Reply:
I remember changing it..... But yeah Im pushing this now. I also return a StatusOK when I dont find the tag to be deleted or there are no tags in the DB